src/Microsoft.VisualStudio.Threading/AsyncLazy`1.cs

```diff
@@ -234,6 +252,69 @@ public T GetValue(CancellationToken cancellationToken)
         }
     }
 
+    /// <summary>
+    /// Disposes of the lazily-initialized value if disposable, and causes all subsequent attempts to obtain the value to fail.
+    /// </summary>
+    /// <remarks>
+    /// <para>Calling this method will put this object into a disposed state where future calls to obtain the value will throw <see cref="ObjectDisposedException"/>.</para>
+    /// <para>If the value has already been produced and implements <see cref="IDisposable"/>, it will be disposed of.
+    /// If the value factory has already started but has not yet completed, its value will be disposed of when the value factory completes.</para>
+    /// <para>If prior calls to obtain the value are in flight when this method is called, those calls <em>may</em> complete and their callers may obtain the value, although <see cref="IDisposable.Dispose"/>
+    /// may have been or will soon be called on the value, leading those users to experience a <see cref="ObjectDisposedException"/>.</para>
+    /// <para>Note all conditions based on the value implementing <see cref="IDisposable"/> is based on the actual value, rather than the <typeparamref name="T"/> type argument.
+    /// This means that although <typeparamref name="T"/> may be <c>IFoo</c> (which does not implement <see cref="IDisposable"/>), the concrete type that implements <c>IFoo</c> may implement <see cref="IDisposable"/>
+    /// and thus be treated as a disposable object as described above.</para>
+    /// </remarks>
+    public void Dispose()
+    {
+        Task<T>? localValueTask = null;
+        IDisposable? localValue = default;
+        lock (this.syncObject)
+        {
+            if (this.value == DisposedSentinel)
+            {
+                return;
+            }
+
+            switch (this.value?.Status)
+            {
+                case TaskStatus.RanToCompletion:
+                    // We'll dispose of the value inline, outside the lock.
+                    localValue = this.value.Result as IDisposable;
+                    break;
+                case TaskStatus.Faulted:
+                case TaskStatus.Canceled:
+                    // Nothing left to do.
+                    break;
+                default:
+                    // We'll schedule the value for disposal outside the lock so it can be synchronous with the value factory,
+                    // but will not execute within our lock.
+                    localValueTask = this.value;
+                    break;
+            }
+
+            // Shut out all future callers from obtaining the value.
+            this.value = DisposedSentinel;
+
+            // Release associated memory.
+            this.joinableTask = null;
+            this.valueFactory = null;
+        }
+
+        if (localValue is not null)
+        {
+            localValue.Dispose();
+        }
+        else if (localValueTask is not null)
+        {
+            localValueTask.ContinueWith(
```

Comment:
The sematic of this Dispose is essentially to ensure the object would be disposed if it is created. It would wait the disposing logic to happen. That is fine, and maybe a desirable behavior in many scenarios, but some clients might need track when the disposing actually done (if it may hold some resources or certain things which might be important under certain circumstances.)  Is it possible this class may also implement DisposeAsync, and potentially provide a way for consumers to wait on the disposing logic to be done.

Reply:
i guess if the user did that, it would leave a race condition that the lazy was triggering on another thread, and it was the reason why you add Dispose method in the first place?  Of course, we can still work around this issue maybe swap a flag or something in the factory method, so it would not be a blocking issue, but just a suggestion.

I don't have direct sample on the top of my head, I can see one reason behind an async version is that the caller has a way to wait the disposing to happen before the process ends, while the current Dispose implementation would not give a chance.

Reply:
I'm adding a `DisposeValueAsync` method, and the `DisposeValue` method will now block on disposal.
Both of these methods only do anything at all (including block) if they are the _first_ call to dispose. Subsequent ones just skip. This is a common disposal pattern, and easier than being JTF-safe in blocking and lighter-weight. Arguably, if only one system wants to block something on disposal (like shutdown), it's enough for any one of the dispose callers to do that block.